File: 프로그래머스/lv1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.py
def solution(dartResult):
    answer = [0]
    
    score = { 'S': '**1', 'D': '**2', 'T': '**3' }
    
    for c in dartResult:
        if c.isdigit():
            answer[-1] = answer[-1] * 10 + int(c)
        elif c in score:
            answer[-1] = eval(f'{answer[-1]}{score[c]}')
            answer.append(0)
        elif c == '*':
            answer[-2] *= 2
            if len(answer) > 2:
                answer[-3] *= 2
        elif c == '#':
            answer[-2] *= -1
            
    return sum(answer)

File: 프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public int solution(String dartResult) {
        List<Integer> answer = new ArrayList<>();
        answer.add(0);

        int len = dartResult.length();
        for (int i = 0; i < len; i++) {
            char c = dartResult.charAt(i);

            if (Character.isDigit(c)) {
                int lastIdx = answer.size() - 1;
                answer.set(lastIdx, answer.get(lastIdx) * 10 + (c - '0'));
            } else if (c == 'S' || c == 'D' || c == 'T') {
                int lastIdx = answer.size() - 1;
                int currentScore = answer.get(lastIdx);
                
                if (c == 'D') {
                    currentScore = currentScore * currentScore;
                } else if (c == 'T') {
                    currentScore = currentScore * currentScore * currentScore;
                }
                
                answer.set(lastIdx, currentScore);
                answer.add(0);
            } else if (c == '*') {
                int size = answer.size();
                answer.set(size - 2, answer.get(size - 2) * 2);
                if (size > 2) {
                    answer.set(size - 3, answer.get(size - 3) * 2);
                }
            } else if (c == '#') {
                int size = answer.size();
                answer.set(size - 2, answer.get(size - 2) * -1);
            }
        }

        int sum = 0;
        for (int score : answer) {
            sum += score;
        }
        return sum;
    }
}